subsys/bluetooth/mesh/friend.c

```diff
@@ -340,33 +313,62 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
 
 	net_buf_add_mem(buf, sdu->data, sdu->len);
 
-	/* We re-encrypt and obfuscate using the received IVI rather than
-	 * the normal TX IVI (which may be different) since the transport
-	 * layer nonce includes the IVI.
-	 */
-	if (bt_mesh_net_encrypt(enc, &buf->b, info->iv_index, false)) {
-		BT_ERR("Re-encrypting failed");
-		goto failed;
+	return buf;
+}
+
+static int encrypt_friend_pdu(struct bt_mesh_friend *frnd, struct net_buf *buf, bool master_cred)
+{
+	const u8_t *enc, *priv;
+	u8_t nid;
+
+	struct bt_mesh_subnet *sub = bt_mesh_subnet_get(frnd->net_idx);
+
+	if (master_cred) {
+		enc = sub->keys[sub->kr_flag].enc;
+		priv = sub->keys[sub->kr_flag].privacy;
+		nid = sub->keys[sub->kr_flag].nid;
+	} else {
+		if (friend_cred_get(sub, frnd->lpn, &nid, &enc, &priv)) {
+			BT_ERR("friend_cred_get failed");
+			return -ENOENT;
+		}
 	}
 
-	if (bt_mesh_net_obfuscate(buf->data, info->iv_index, priv)) {
-		BT_ERR("Re-obfuscating failed");
-		goto failed;
+	u16_t src = ((buf->data[5] << 16) | (buf->data[6]));
```

Comment:
`src = sys_get_be16(&buf->data[5]);`

Also, please move the variable declarations to the beginning of the scope (the function in this case). I don't think this is enforced by any Zephyr global coding style, but that's what we've followed in the stack. One benefit is that it's easier to assess the stack footprint of the function that way.